packages/vgrammar/src/transforms/data/bin.ts

```diff
@@ -0,0 +1,23 @@
+import { isValid } from '@visactor/vutils';
+import type { BinTransformOption } from '../../types';
+
+export const transform = (options: BinTransformOption, upstreamData: any[]) => {
+  if (!upstreamData || upstreamData.length === 0) {
+    return upstreamData;
+  }
+
+  const { extent, bins, step } = options;
+  const range = extent[1] - extent[0];
+  const binStep = isValid(step) ? step : range / (bins ?? 10);
+
+  upstreamData.forEach(datum => {
```

Comment:
it's better to return a new array by default not modify the original data?
many other transforms have the same problem

Reply:
Fixed.